axelrod/strategies/dbs.py

```diff
@@ -0,0 +1,354 @@
+            self.Pi.update(self.Rc)
+            self.Pi.update(self.Rd)
+            self.Pi.update(Rp)
+
+        # React to the opponent's last move
+        return MoveGen((self.history[-1],opponent.history[-1]),self.Pi,depth_search_tree=self.tree_depth)
+
+
+# Policy as defined in the article, i.e. a set of (last_move,p) where p is
+# the probability to cooperate in the next move considering last move
+class Policy(dict):
+
+    def __init__(self):
+        super().__init__()
+
+    @classmethod
+    def prob_policy(cls,pCC,pCD,pDC,pDD):
+        pol = cls()
+        pol[(C,C)]=pCC
+        pol[(C,D)]=pCD
+        pol[(D,C)]=pDC
+        pol[(D,D)]=pDD
+        return pol
+    
+    def proba(self,action1,action2):
+        return self[(action1,action2)]
+
+# Nodes used to build a tree for the tree-search procedure
+# The tree has Determinist ans Stochastic nodes, as the opponent's
+# strategy is learned as a probability distribution
+
+class Node(object):
+
+    # abstract method
+    def get_siblings(self):
+        raise NotImplementedError('subclasses must override get_siblings()!')
+
+    # abstract method
+    def is_stochastic(self):
+        raise NotImplementedError('subclasses must override is_stochastic()!')
+
+class StochasticNode(Node):
+    "Node that have a probability p to get to each sibling"
+    "Nodes (C,*) or (D,*)"
+
+    def __init__(self,own_action,pC,depth):
+        self.pC = pC
+        self.depth = depth
+        self.own_action = own_action
+
+    def get_siblings(self):
+        # siblings of a stochastic node get depth += 1 
+        opponent_c_choice = DeterministNode(self.own_action,C,self.depth+1)
+        opponent_d_choice = DeterministNode(self.own_action,D,self.depth+1)
+        return (opponent_c_choice,opponent_d_choice)
+
+    def is_stochastic(self):
+        return True
+
+class DeterministNode(Node):
+    "Node (C,C),(C,D),(D,C), or (D,D) with determinist choice for siblings"
+
+    def __init__(self,action1,action2,depth):
+        self.action1 = action1
+        self.action2 = action2
+        self.depth = depth
+
+    def get_siblings(self,policy):
+        # build 2 siblings (C,*) and (D,*)
+        # siblings of a DeterministicNode are Stochastic, and are of the
+        # same depth 
+        c_choice = StochasticNode(C,
+                policy.proba(self.action1,self.action2),self.depth)
+        d_choice = StochasticNode(D,
+                policy.proba(self.action1,self.action2),self.depth)
+        return (c_choice,d_choice)
+
+    def is_stochastic(self):
+        return False
+
+    def get_value(self):
+        if (self.action1==C and self.action2==C):
+            return 3
+        elif (self.action1==C and self.action2==D):
+            return 0
+        elif (self.action1==D and self.action2==C):
+            return 5
+        elif (self.action1==D and self.action2==D):
+            return 1
+
+# tree search function (minimax search procedure)
+def F(begin_node,policy,max_depth):
```

Comment:
Let's give `F` a good name. If it's called `F` in the reference then let's say so in the doc string